Transfer CalcJob (#4194)
aiidateam/aiida-core: docs/source/howto/data.rst

You can use this feature to get files from a remote machine to be saved in a local :py:class:`~aiida.orm.nodes.data.folder.FolderData`.
The specifications of what to copy are provided through an input of type

.. code-block:: ipython

    In [1]: instructions_cont = {}
        ... instructions_cont['retrieve_files'] = True
        ... instructions_cont['symlink_files'] = [
        ...     ('node_keyname', 'source/path/filename', 'target/path/filename'),
        ... ]
        ... instructions_node = orm.Dict( dict = instructions_cont )

The ``'source/path/filename'`` and ``'target/path/filename'`` are both relative paths (to their respective folders).
The ``node_keyname`` is a string that will be used when providing the source :py:class:`~aiida.orm.nodes.data.remote.RemoteData` node to the calculation.
You also need to provide the computer between which the transfer will ocurr:

.. code-block:: ipython

    In [2]: transfer_builder = CalculationFactory('core.transfer').get_builder()
        ... transfer_builder.instructions = instructions_node
        ... transfer_builder.source_nodes = {'node_keyname': source_node}

>         ... transfer_builder.metadata.computer = computer_node

Reviewer: a `Computer` instance is not really a node. It doesn't live in the node table and also is not a subclass of `Node`. You also don't explain how to get it.
```suggestion
        ... transfer_builder.metadata.computer = source_node.computer
```

Author: So, the thing is that this fix works for when the `source_node` is of type `RemoteData` (which is the case for the first example). However, in the second example I say they can do the same thing but just changing the `dict` setup, so if they copy this for the second part it won't work. There they have to load the remote computer they want to use and pre-set it themselves, I cannot give further details of which one they want to use. I can remove the `_node` part though.

Reviewer: I would actually keep the change I suggested and then simply explain what you just did in the example with the `FolderData` source node. There I would then suggest to show them to use `load_computer('some-computer-label')`. This is the typical way to load a `Computer` instance through the API. You can make it as explicit as the following for clarity
```
transfer_builder.metadata.computer = load_computer('some-computer-label')
```

> The variables ``source_node`` and ``computer_node`` must already contain the corresponding nodes.

Reviewer: ```suggestion
The variable ``source_node`` here corresponds to the ``RemoteData`` node whose contents need to be retrieved.
```

Finally, you just run or submit the calculation as you would do with any other:

.. code-block:: ipython

    In [2]: from aiida.engine import submit

>         ... submit( transfer_builder )

Reviewer: ```suggestion
        ... submit(transfer_builder)
```

You can also use this to copy local files into a new :py:class:`~aiida.orm.nodes.data.remote.RemoteData` folder.
In this case the ``source_node`` must be of type :py:class:`~aiida.orm.nodes.data.folder.FolderData`, and you must adapt the instructions to set ``'retrieve_files'`` to ``False`` and use a ``'local_files'`` list instead of the ``'symlink_files'``:

.. code-block:: ipython

    In [1]: instructions_cont = {}
        ... instructions_cont['retrieve_files'] = False
        ... instructions_cont['local_files'] = [
        ...     ('node_keyname', 'source/path/filename', 'target/path/filename'),
        ... ]

>         ... instructions_node = orm.Dict( dict = instructions_cont )

Reviewer: ```suggestion
        ... instructions_node = orm.Dict(dict=instructions_cont)
```

> In both cases, you can copy many files by appending them to the respective list.

Reviewer: ```suggestion
Both when uploading or retrieving, you can copy multiple files by appending them to the list of the ``local_files`` or ``symlink_files`` keys in the instructions input, respectively.
```
